core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -753,39 +753,63 @@ class ReplicaManager(val config: KafkaConfig,
         val localProduceResults = appendToLocalLog(internalTopicsAllowed = internalTopicsAllowed,
           origin, verifiedEntries, requiredAcks, requestLocal, verificationGuards.toMap)
         debug("Produce to local log in %d ms".format(time.milliseconds - sTime))
-        
-        val unverifiedResults = unverifiedEntries.map { case (topicPartition, error) =>
-          val message = if (error == Errors.INVALID_TXN_STATE) "Partition was not added to the transaction" else error.message()
-          topicPartition -> LogAppendResult(
-            LogAppendInfo.UNKNOWN_LOG_APPEND_INFO,
-            Some(error.exception(message))
-          )
-        }
 
-        val errorResults = errorsPerPartition.map { case (topicPartition, error) =>
-          topicPartition -> LogAppendResult(
-            LogAppendInfo.UNKNOWN_LOG_APPEND_INFO,
-            Some(error.exception())
-          )
+        def produceStatusResult(appendResult: Map[TopicPartition, LogAppendResult],
+                                useCustomMessage: Boolean): Map[TopicPartition, ProducePartitionStatus] = {
+          appendResult.map { case (topicPartition, result) =>
+            topicPartition -> ProducePartitionStatus(
+              result.info.lastOffset + 1, // required offset
+              new PartitionResponse(
+                result.error,
+                result.info.firstOffset.map[Long](_.messageOffset).orElse(-1L),
+                result.info.lastOffset,
+                result.info.logAppendTime,
+                result.info.logStartOffset,
+                result.info.recordErrors,
+                if (useCustomMessage) result.exception.get.getMessage else result.info.errorMessage
```

Comment:
I'm not sure I understand. Previously the result.info.errorMessage was null and it will continue to be so for all error partitions. LogAppendInfo.UnknownLogAppendInfo will always have a null error.

I thought for consistency I should not change this.

Reply:
I suppose that you are saying that `result.info.errorMessage` is always null when `useCustomMessage` is true because we know that `LogAppendInfo.UnknownLogAppendInfo` does not set it. Am I right?

What I was wondering is why don't we set it all the time. I suppose that what you did here is fine.

Reply:
Correct -- when we use custom message, we have always used LogAppendInfo.UnknownLogAppendInfo.

I'm not sure why it is this way. Perhaps if it is the default error message (which was the case previously) we didn't want to bloat the response.